### ISSUE TITLE
Consistency improvements to epoch and counter fields

### ISSUE THREAD

**Issue by bgirardeau-figma** [osquery/osquery#7799]
# Blueprint

**tl;dr** Make epoch and counter behavior more consistent to support both periodic snapshots and stream alerting from the same differential rule

## Problem
The schedule epoch and schedule counter fields described [here](https://osquery.readthedocs.io/en/stable/deployment/logging/) provide a way to get periodic snapshots of the current state of a host, while still leveraging the efficiency and low latency of using differential queries.

However if the epoch is changed regularly, it isn't possible to reliably determine whether log entries are new changes or previously seen changes that are part of the periodic snapshot. From the documentation, it suggests ignoring events with a counter of 0 as part of an "initial run" to solve this issue, but there are some inconsistencies that keep this from working well:
* New changes between the last run in a previous epoch and first run in the next would only appear with counter=0
* Evented tables (under default optimization mode) never produce snapshots but do use counter=0 for some events
* There is no way to distinguish between the actual "initial run" of osquery, and the periodic snapshots generated later

This prevents using a changing epoch if the rule is used in stream based alerting workflows (without either potentially missing data or having repeated alerts). Instead users have to use a separate snapshot rule to get periodic snapshots, which is more overhead to manage and more difficult to analyze together with the separate differential rule data.

## Proposal
With a few changes, we can make the meaning of epoch and counter more consistent to support the combined snapshot / stream based alerting use case.

### 1. Counter=0 is reserved for periodic snapshots
If a run does not produce a snapshot (notably for an evented table), it will not use counter=0 -- instead its counter will start at 1. This means counter=0 is reserved for periodic snapshots, so it can be safely ignored for use cases that only examine non-snapshot data.

### 2. Ensure periodic snapshots don't contain new changes
When a new epoch is encountered and a periodic snapshot will be logged, the client will first calculate the differential between the snapshot and the last run of the previous epoch. If there are differences, they are logged before the snapshot as if the changes were part of the previous epoch.

This means that clients that ignore periodic snapshots still reliably see all new changes. The exception is for the first run ever of a rule, which has no previous epoch / run to compare with, which is addressed by the next change.

### 3. New `previous_epoch` field distinguishes between first snapshot and later snapshots
The `previous_epoch` field will have the epoch of the previous run of the rule, or 0 if this is the first time it has run. This allows for distinguishing between the first run/snapshot of the rule and periodic snapshots in future epoch.

As an alternative, a boolean top level `is_first_run` field would be enough for this particular use case. Using `previous_epoch` seems nicer to also provide more useful data for consistency checks and querying generally:
* users can more easily compare between successive snapshots by joining on `previous_epoch = epoch`
* users can monitor if `last snapshot + differential changes = next snapshot` as a consistency check

### 4. Log empty periodic snapshot if using batch format (?)
When the epoch changes or on first run, if the periodic snapshot is empty, then an empty batch result would be logged. Currently no log item is produced in this case, which makes it hard to distinguish between the rule not running and no results being returned. This is unavoidable in the event stream format, at least without a new event type to indicate it, but it seems reasonable to have a batch log entry with empty lists to indicate this case consistently.

There is a reasonable workaround to instead look for epochs with missing "counter=0", so it's not as important to do this if we think it might cause confusion/issues.

**Comment by bgirardeau-figma** [comment]
The referenced PRs implement proposal items [(1)](https://github.com/osquery/osquery/pull/7801) and [(4)](https://github.com/osquery/osquery/pull/7803).

Implementing proposal items (2) and (3) was easiest to do on top of those PRs, so draft code is here in the meantime:
* (3) `previous_epoch` field: https://github.com/bgirardeau-figma/osquery/commit/ad508b523232810f17d0b2f5b58aceda6dde69e6
* (2) periodic snapshots don't contain new changes: https://github.com/bgirardeau-figma/osquery/commit/34b0312c37fe943d5447f506ce3c3f8cbe208a83
